piptools/build.py

```diff
@@ -0,0 +1,174 @@
+
+PYPROJECT_TOML = "pyproject.toml"
+
+_T = TypeVar("_T")
+
+
+if sys.version_info >= (3, 10):
+    from importlib.metadata import PackageMetadata
+else:
+
+    class PackageMetadata(Protocol):
+        @overload
+        def get_all(self, name: str, failobj: None = None) -> list[Any] | None:
+            ...
+
+        @overload
+        def get_all(self, name: str, failobj: _T) -> list[Any] | _T:
+            ...
+
+
+@dataclass
+class ProjectMetadata:
+    extras: tuple[str, ...]
+    requirements: tuple[InstallRequirement, ...]
+    build_requirements: tuple[InstallRequirement, ...]
+
+
+def build_project_metadata(
+    src_file: pathlib.Path,
+    build_targets: tuple[str, ...],
+    *,
+    isolated: bool,
+    quiet: bool,
+) -> ProjectMetadata:
+    """
+    Return the metadata for a project.
+
+    Uses the ``prepare_metadata_for_build_wheel`` hook for the wheel metadata
+    if available, otherwise ``build_wheel``.
+
+    Uses the ``prepare_metadata_for_build_{target}`` hook for each ``build_targets``
+    if available.
+
+    :param src_file: Project source file
+    :param build_targets: A tuple of build targets to get the dependencies
+                                of (``sdist`` or ``wheel`` or ``editable``).
+    :param isolated: Whether to run invoke the backend in the current
+                     environment or to create an isolated one and invoke it
+                     there.
+    :param quiet: Whether to suppress the output of subprocesses.
+    """
+
+    src_dir = src_file.parent
+    with _create_project_builder(src_dir, isolated=isolated, quiet=quiet) as builder:
+        metadata = _build_project_wheel_metadata(builder)
+        extras = tuple(metadata.get_all("Provides-Extra") or ())
+        requirements = tuple(
+            _prepare_requirements(metadata=metadata, src_file=src_file)
+        )
+        build_requirements = tuple(
+            _prepare_build_requirements(
+                builder=builder,
+                src_file=src_file,
+                build_targets=build_targets,
+                package_name=_get_name(metadata),
+            )
+        )
+        return ProjectMetadata(
+            extras=extras,
+            requirements=requirements,
+            build_requirements=build_requirements,
+        )
+
+
+@contextlib.contextmanager
+def _create_project_builder(
+    src_dir: pathlib.Path, *, isolated: bool, quiet: bool
+) -> Iterator[build.ProjectBuilder]:
+    if quiet:
+        runner = pyproject_hooks.quiet_subprocess_runner
+    else:
+        runner = pyproject_hooks.default_subprocess_runner
+
+    if not isolated:
+        yield build.ProjectBuilder(src_dir, runner=runner)
+        return
+
+    with build.env.DefaultIsolatedEnv() as env:
+        builder = build.ProjectBuilder.from_isolated_env(env, src_dir, runner)
+        env.install(builder.build_system_requires)
+        env.install(builder.get_requires_for_build("wheel"))
```

Comment:
@apljungquist looks like these installs have to be called with `os.environ['PIP_CONSTRAINT']` set to a generated file with `-P` values dumped in order to install the backend deps we desire.

Reply:
@webknjaz Checking I understand this comment as being related to [this note](https://github.com/jazzband/pip-tools/blob/53309647980e2a4981db54c0033f98c61142de0b/README.md?plain=1#L580) in the README: "- `pip` must be used with the `PIP_CONSTRAINT` environment variable to lock dependencies in build environments as documented in [#8439](https://github.com/pypa/pip/issues/8439)."

Rather than just installing the requirements as listed in the metadata, you're suggesting that:

1. The current value of `os.environ["PIP_CONSTRAINT"] should be saved to avoid irrevocably clobbering it
2. For each `env.install` step, in addition to the requirements being passed explicitly, they also need to be added to a temporary file referenced from `PIP_CONSTRAINT` to force the dependencies to actually be respected
3. Once the installs are done, `PIP_CONSTRAINT` should be reverted back to its original value.

To try to get CI going, I'm going to try just setting `PIP_CONSTRAINT` in the failing test case, which would mean this discovery still needs its own bug report.

Reply:
Attempted workaround in https://github.com/jazzband/pip-tools/pull/2105/commits/bc3f17bf0c06ba7d7d44783aca7ec817870c0e18 still didn't get the affected test case passing when testing locally, but I'm trying it in CI anyway in case the local failure was specific to Python 3.12.

Reply:
I tried a different workaround, specifying a known version of setuptools in the test metadata: https://github.com/jazzband/pip-tools/pull/2105/commits/c2bd925433a144f88d0ebbfb31c4648349387d6e

Reply:
Even with both attempted workarounds combined, the build dependency compilation test still fails: https://github.com/jazzband/pip-tools/pull/2105/files#diff-99b06d953a9978ee564df2e67547b8a76f3b1ab116ab7c89a9005b9239f2a904

This is with setuptools 70.0.0 explicitly requested in the test project config, and with a constraints file constraining both `setuptools` and `wheel` set via the `pip-tools` CLI option *and* via the `PIP_CONSTRAINT` environment variable.

Reply:
@ncoghlan

> 2. For each `env.install` step, in addition to the requirements being passed explicitly, they also need to be added to a temporary file referenced from `PIP_CONSTRAINT` to force the dependencies to actually be respected

I think that the `[build-system]requires` value is there already as an explicit install request. But the constraints coming from other places like the CLI command or other/additional input files are not taken into account. I think they should be dumped into a temporary file and that file would need to be set in `PIP_CONSTRAINT` around such invocations.

Yes to points (1) and (3), though.

Reply:
@webknjaz Your suggestion to set the `PIP_CONSTRAINT` environment variable appears to work, but another problem surfaced; the failing test in question doesn't use build isolation (it succeeds when that is explicitly requested) – what do you suggest should be done in such a case?

Reply:
@chrysle oh, I missed this. Could you clarify what you mean by requesting it? Are you talking about calling some tooling outside the test suite or making said change in the tests? Perhaps a PR would make it more obvious what you mean.

Reply:
No problem! I was referring to the `--build-isolation` boolean flag. The failing test explicitly turned that off: 

https://github.com/jazzband/pip-tools/blob/53309647980e2a4981db54c0033f98c61142de0b/tests/test_cli_compile.py#L3438

However, this means that `_create_project_builder` yields `build.ProjectBuilder` relying on the outside environment and the (latest) version of `setuptools` that was installed together with `pip-tools` - no doubt to improve performance - but the test naturally fails.

https://github.com/jazzband/pip-tools/blob/53309647980e2a4981db54c0033f98c61142de0b/piptools/build.py#L192-L194

Using build isolation makes the test pass after setting `PIP_CONSTRAINT` as you suggested.

Reply:
@chrysle thanks for the explanation! Yes, you're right that this was a trade-off made in the name of performance. It sounds like we need a separate regression/acceptance test checking this corner case in isolation. It'll be slower but, hopefully, one test won't make a big difference.